app/health_check.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type PluginHealthCheckJob struct {
+	server    *Server
+	task      *model.ScheduledTask
+	taskMutex sync.Mutex
+}
+
+const (
+	PLUGIN_HEALTH_CHECK_TASK_NAME = "Plugin Health Check"
+	PLUGIN_HEALTH_CHECK_INTERVAL  = 5 // seconds
+	HEALTH_CHECK_FAIL_LIMIT       = 3
+	ENABLE_HEALTH_CHECK_JOB       = false
+)
+
+func (s *Server) InitPluginHealthCheckJob() {
+	if ENABLE_HEALTH_CHECK_JOB {
```

Comment:
I imagine this constant is just part of the WIP bits and this would instead be a configurable setting?

Reply:
Right. This was set here as an ease for initial implementation. This push was mainly meant to gather more in-depth requirements and generate discussion within the team.

As for this specific suggestion, are you thinking that the configuration for fields such as backoff delay, health check interval, and restart limit should be configured per plugin? If it is per plugin, the configuration could be stored in the specific plugin's `plugin.json` file, with defaults stored in the server's `config.json` file. For the "global" config of these health check settings, it could be stored in the server's `config.json` like so:

```
{
  "PluginSettings": {
    "HealthCheckSettings": {
      "HealthCheckInterval": 30,
      "HealthCheckFailLimit": 3 (how many failed checks before we should restart, excluding short-circuiting fatal checks like process does not exist),
      "HealthCheckRestartLimit": 3 (how many restarts before we say that this plugin will continue to fail),
      "HealthCheckRestartWaitTime": 60 (how long we wait for plugin to be considered healthy before we decide to restart again, because we don't want to restart again if it fails on startup),
      "EnableHealthCheckJob": true,
      "EnableHealthChecksInAdminConsole": true
    }
  }
}

```
There are definitely more configuration options we could include, and some listed above may not be included or have overlapping use cases with other config options.

Reply:
Actually ideally I think I'd rather not expose it as configuration settings to the admin or plugin authors at all and just take care of it ourselves. I don't think admins or plugin authors would better be able to set these than us, so I think this is a spot where we should be opinionated and just pick a set strategy